tests/testutils.py

```diff
@@ -24,15 +26,56 @@
 from twisted.web.http_headers import Headers
 from twisted.web.server import Request
 
+import psycopg2
+
 from sygnal.http import PushGatewayApiServer
 from sygnal.sygnal import CONFIG_DEFAULTS, Sygnal, merge_left_with_defaults
 
 REQ_PATH = b"/_matrix/push/v1/notify"
 
+# the dbname we will connect to in order to create the base database.
+POSTGRES_DBNAME_FOR_INITIAL_CREATE = "postgres"
+POSTGRES_USER = environ.get("TEST_POSTGRES_USER", None)
+POSTGRES_PASSWORD = environ.get("TEST_POSTGRES_PASSWORD", None)
+POSTGRES_HOST = environ.get("TEST_POSTGRES_HOST", None)
+
 
 class TestCase(unittest.TestCase):
     def config_setup(self, config):
-        config["db"]["dbfile"] = ":memory:"
+        self.dbname = "_sygnal_%s" % (time_ns())
+        config["db"] = {
+            "user": POSTGRES_USER,
+            "password": POSTGRES_PASSWORD,
+            "database": self.dbname,
+            "host": POSTGRES_HOST,
+        }
+
+    def _set_up_database(self, dbname):
+        conn = psycopg2.connect(
+            database=POSTGRES_DBNAME_FOR_INITIAL_CREATE,
+            user=POSTGRES_USER,
+            password=POSTGRES_PASSWORD,
+            host=POSTGRES_HOST,
+        )
+        conn.autocommit = True
+        cur = conn.cursor()
+        cur.execute("DROP DATABASE IF EXISTS %s;" % (dbname,))
+        cur.execute("CREATE DATABASE %s;" % (dbname,))
```

Comment:
You need to `conn.commit()` before closing.

Reply:
Not in this case. We're using `conn.autocommit` like in https://github.com/matrix-org/synapse/blob/d9239b52571c7e07fa9b035de1ad46bb602ed163/tests/utils.py#L68-L79 because running a DROP and a CREATE inside a single txn isn't allowed.

Reply:
Oh! Missed that, interesting